include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,819 @@
+	 *
+	 * Value 0 mean implementation use its default value.
+	 *
+	 */
+	uint32_t max_level;
+
+	/* Supported hash algorithms */
+	odp_comp_hash_algos_t hash_alg;
+} odp_comp_alg_capability_t;
+
+/**
+ * Comp API dictionary type
+ *
+ */
+typedef struct odp_comp_dict_t {
+	/** pointer to character array */
+	uint8_t *buf;
+	/** length of the dictionary. */
+	uint32_t len;
+} odp_comp_dict_t;
+
+/**
+ * Comp API algorithm specific parameters
+ *
+ */
+typedef struct odp_comp_alg_param_t {
+	struct comp_alg_def_param {
+		/** compression level where
+		 * ODP_COMP_LEVEL_MIN <= level <= ODP_COMP_LEVEL_MAX
+		 */
+		odp_comp_level_t level;
+		/** huffman code to use */
+		odp_comp_huffman_code_t comp_code;
+	} deflate;
+	struct comp_alg_zlib_param {
+			/** deflate algo params */
+			struct comp_alg_def_param def;
+	} zlib;
+} odp_comp_alg_param_t;
+
+/**
+ * Comp API data range specifier
+ *
+ */
+typedef union odp_comp_data_t {
+	struct {
+		/** packet */
+		odp_packet_t packet;
+
+		/** packet data range to operate on  */
+		odp_packet_data_range_t data_range;
+	} pkt;
+} odp_comp_data_t;
+
+ /**
+ * Comp API session creation parameters
+ *
+ */
+typedef struct odp_comp_session_param_t {
+	/** Compress vs. Decompress operation */
+	odp_comp_op_t op;
+
+	/** Sync vs Async
+	 *
+	 * When mode = ODP_COMP_SYNC, odp_comp_compress()/odp_comp_decomp()
+	 * should be called.
+	 *
+	 * When mode = ODP_COMP_ASYNC, odp_comp_compress_enq()/
+	 * odp_comp_decomp_enq() should be called.
+	 *
+	 * Use odp_comp_capability() for supported mode.
+	 *
+	 */
+	odp_comp_op_mode_t mode;
+
+	/** Compression algorithm
+	 *
+	 *  Use odp_comp_capability() for supported algorithms.
+	 */
+	odp_comp_alg_t comp_alg;
+
+	/** Hash algorithm
+	 *
+	 *  Use odp_comp_alg_capability() for supported hash algo for
+	 *  compression algo given as comp_alg. Implementation should not
+	 *  support hash only operation on data. output should always contain
+	 *  data + hash.
+	 *
+	 */
+	odp_comp_hash_alg_t hash_alg;
+
+	/** parameters specific to compression */
+	odp_comp_alg_param_t alg_param;
+
+	/** Async mode completion event queue
+	 *
+	 * When mode = ODP_COMP_ASYNC, user should wait on ODP_EVENT_PACKET
+	 * with subtype ODP_EVENT_PACKET_COMP on this queue.
+	 *
+	 * By default, implementation enques completion events in-order-of
+	 * request submission and thus queue is considered ordered.
+	 *
+	 * Please note, behavior could be changed or enhanced
+	 * to queue event in-order-of their completion to enable
+	 * performance-oriented application to leverage hw offered parallelism.
+	 * However, this will be subject to application requirement and more
+	 * explicit defined use-case.
+	 *
+	 */
+	odp_queue_t compl_queue;
+} odp_comp_session_param_t;
+
+/**
+ * Comp API operation parameters.
+ * Called to process each data unit.
+ *
+ */
+typedef struct odp_comp_op_param_t {
+	/** Session handle from creation */
+	odp_comp_session_t session;
+
+	/** User context */
+	void *ctx;
```

Comment:
its is still useful to allow user to tie any compression specific data in to it user context ..ex. writing this with compression session handles in case of multiple async calls to identify which packet belong to which session OR say compression/decompression request numbers.